src/util/sighash.rs

```diff
@@ -0,0 +1,909 @@
+                .consensus_encode(&mut writer)?;
+            txin.sequence.consensus_encode(&mut writer)?;
+        } else {
+            (input_index as u32).consensus_encode(&mut writer)?;
+        }
+
+        // If an annex is present (the lowest bit of spend_type is set):
+        //      sha_annex (32): the SHA256 of (compact_size(size of annex) || annex), where annex
+        //      includes the mandatory 0x50 prefix.
+        if let Some(annex) = annex {
+            let mut enc = sha256::Hash::engine();
+            annex.consensus_encode(&mut enc)?;
+            let hash = sha256::Hash::from_engine(enc);
+            hash.consensus_encode(&mut writer)?;
+        }
+
+        // * Data about this output:
+        // If hash_type & 3 equals SIGHASH_SINGLE:
+        //      sha_single_output (32): the SHA256 of the corresponding output in CTxOut format.
+        if sighash == SigHashType::Single {
+            let mut enc = sha256::Hash::engine();
+            self.tx
+                .output
+                .get(input_index)
+                .ok_or_else(|| Error::SingleWithoutCorrespondingOutput {
+                    index: input_index,
+                    outputs_size: self.tx.output.len(),
+                })?
+                .consensus_encode(&mut enc)?;
+            let hash = sha256::Hash::from_engine(enc);
+            hash.consensus_encode(&mut writer)?;
+        }
+
+        //     if (scriptpath):
+        //         ss += TaggedHash("TapLeaf", bytes([leaf_ver]) + ser_string(script))
+        //         ss += bytes([0])
+        //         ss += struct.pack("<i", codeseparator_pos)
+        if let Some(ScriptPath {
+            script,
+            leaf_version,
+            code_separator_pos,
+        }) = script_path
+        {
+            let mut enc = TapLeafHash::engine();
+            leaf_version.consensus_encode(&mut enc)?;
+            script.consensus_encode(&mut enc)?;
+            let hash = TapLeafHash::from_engine(enc);
+
+            hash.into_inner().consensus_encode(&mut writer)?;
+            KEY_VERSION_0.consensus_encode(&mut writer)?;
+            code_separator_pos.consensus_encode(&mut writer)?;
+        }
+
+        Ok(())
+    }
+
+    /// Compute the BIP341 sighash for any flag type.
+    pub fn taproot_signature_hash(
+        &mut self,
+        input_index: usize,
+        prevouts: &Prevouts,
+        annex: Option<Annex>,
+        script_path: Option<ScriptPath>,
+        sighash_type: SigHashType,
+    ) -> Result<TapSighashHash, Error> {
+        let mut enc = TapSighashHash::engine();
+        self.taproot_encode_signing_data_to(
+            &mut enc,
+            input_index,
+            prevouts,
+            annex,
+            script_path,
+            sighash_type,
+        )?;
+        Ok(TapSighashHash::from_engine(enc))
+    }
+
+    /// Encode the BIP143 signing data for any flag type into a given object implementing a
+    /// [`std::io::Write`] trait.
```

Comment:
Nit:

```suggestion
    /// [`io::Write`] trait.
```

as was specified above (to support no_std)